### PR TITLE
Add support for i128 and u128

### LINKED ISSUE #2822: Implement WASM ABI conversion for `u128` and `i128`
### Motivation

To cover all the integers in the Rust standard library, we should have an implementation for these two missing types `u128` and `i128`. We want to implement at least the `IntoWasmAbi` and `FromWasmAbi` traits.

### Proposed Solution

Since `u64` already implements this with the `Wasm64` ABI type, we can use something like

```rust
#[repr(C)]
pub struct Wasm128 {
    pub low: Wasm64,
    pub high: Wasm64,
}
```

and implement the `FromWasmAbi` and `IntoWasmAbi` traits as follows:

```rust
impl WasmAbi for Wasm128 {}

impl IntoWasmAbi for u128 {
    type Abi = Wasm128;

    #[inline]
    fn into_abi(self) -> Wasm128 {
        Wasm128 {
            low: (self as u64).into_abi(),
            high: ((self >> 64) as u64).into_abi(),
        }
    }
}

impl FromWasmAbi for u128 {
    type Abi = Wasm128;

    #[inline]
    unsafe fn from_abi(js: Wasm128) -> u128 {
        u128::from(u64::from_abi(js.low)) | (u128::from(u64::from_abi(js.high)) << 64)
    }
}
```

We can use a macro like the existing implementation to cover both the `u128` and `i128` cases.

### Alternatives

An alternative is to define a wrapper type around `u128` which I can use in my own crate which I then attach the `wasm_bindgen` macro to. This is undesirable if we can just put this directly in this library.

### PR DESCRIPTION
**Pull request by RunDevelopment**
Fixes #2822

This adds WASM ABI support for `i128` and `u128`. Both are split into 2 `u64` and passed as such in the ABI. 

The following types can now be used in argument and return positions: `i128`, `u128`, `Option<i128>`, and `Option<u128>`. Return positions can also use `Result<i128, E>` and `Result<u128, E>`.

It's important to note that `Result<Option<i128>, E>` is **not** supported. This is because the `WasmAbi` trait supports at most 4 primitives, but `Result<Option<i128>, E>` would require 5 primitives (2 for `i128`, 1 for the option tag, 1 for the result tag, and 1 for the error variant value). Future `Result<Option<T>, E>` ABI optimizations (either combining the result and option tag into one primitive, or raising the number of supported primitives to 5) will make `Result<Option<i128>, E>` possible, but this should be a separate PR IMO.

---

Related to #4201.

### CODE REVIEW

File: crates/cli-support/src/js/binding.rs [hunk 1]
```diff
@@ -982,11 +1029,7 @@ fn instruction(
             js.push(format!("!isLikeNone({0})", val));
             js.push(format!(
                 "isLikeNone({val}) ? {zero} : {val}",
-                zero = if *ty == ValType::I64 {
-                    "BigInt(0)"
-                } else {
-                    "0"
-                }
+                zero = if *ty == ValType::I64 { "0n" } else { "0" }
```
Comment (MattiasBuelens) [inline crates/cli-support/src/js/binding.rs:1032]: This might regress on #3164. We may want to avoid bigint literals everywhere else too.

Comment (RunDevelopment) [reply crates/cli-support/src/js/binding.rs:1032]: Good catch. Thank you! This is definitely a blocker.

After looking at #3164 and the error it intended to fix, I now believe that #3164 shouldn't have been accepted. I'll open an issue about where I'll explain my reasoning in detail.

**Comment by daxpedda** [comment]
Setting on draft until #4246 is resolved.

**Comment by RunDevelopment** [comment]
> Converting to draft until #4246 is resolved.

Either that, or I just use `BigInt(value)` everywhere for now. If we decide to bigint literals in #4246 later, I can just make another PR to change it later. Grepping for `BigInt(` isn't hard.

**Comment by daxpedda** [comment]
> > Converting to draft until #4246 is resolved.
> 
> Either that, or I just use `BigInt(value)` everywhere for now. If we decide to bigint literals in #4246 later, I can just make another PR to change it later. Grepping for `BigInt(` isn't hard.

Yep, please go ahead and just use the constructor!

File: crates/cli-support/src/js/binding.rs (hunk 1 above)
Comment (RunDevelopment) [reply crates/cli-support/src/js/binding.rs]: I'll just use `BigInt` constructors for now.

File: tests/wasm/optional_primitives.js [hunk 2]
```diff
@@ -102,6 +104,22 @@ exports.js_works = () => {
...
+    assert.strictEqual(wasm.optional_i128_identity(wasm.optional_i128_min()), -170141183460469231731687303715884105728n);
+    assert.strictEqual(wasm.optional_i128_identity(wasm.optional_i128_max()), 170141183460469231731687303715884105727n);
+    assert.strictEqual(wasm.optional_i128_identity(0x3_1415_9265_3598_7932_3846n), 0x3_1415_9265_3598_7932_3846n);
+    assert.strictEqual(wasm.optional_i128_identity(-0x3_1415_9265_3598_7932_3846n), -0x3_1415_9265_3598_7932_3846n);
+
+    assert.strictEqual(wasm.optional_u128_identity(wasm.optional_u128_none()), undefined);
+    assert.strictEqual(wasm.optional_u128_identity(wasm.optional_u128_zero()), BigInt('0'));
+    assert.strictEqual(wasm.optional_u128_identity(wasm.optional_u128_one()), BigInt('1'));
+    assert.strictEqual(wasm.optional_u128_identity(wasm.optional_u128_min()), BigInt('0'));
+    assert.strictEqual(wasm.optional_u128_identity(wasm.optional_u128_max()), 340282366920938463463374607431768211455n);
```
Comment (daxpedda) [inline tests/wasm/optional_primitives.js:120]:
```suggestion
    assert.strictEqual(wasm.optional_u128_identity(wasm.optional_u128_max()), 340282366920938463463374607431768211455n);
```
Can't we just use hexadecimal representation here and in other similar ones as well?

File: tests/wasm/optional_primitives.js [hunk 3]
```diff
@@ -102,6 +104,22 @@ exports.js_works = () => {
     assert.strictEqual(wasm.optional_u64_identity(wasm.optional_u64_min()), BigInt('0'));
     assert.strictEqual(wasm.optional_u64_identity(wasm.optional_u64_max()), BigInt('18446744073709551615'));
 
+    assert.strictEqual(wasm.optional_i128_identity(wasm.optional_i128_none()), undefined);
+    assert.strictEqual(wasm.optional_i128_identity(wasm.optional_i128_zero()), BigInt('0'));
+    assert.strictEqual(wasm.optional_i128_identity(wasm.optional_i128_one()), BigInt('1'));
+    assert.strictEqual(wasm.optional_i128_identity(wasm.optional_i128_neg_one()), BigInt('-1'));
+    assert.strictEqual(wasm.optional_i128_identity(wasm.optional_i128_min()), -170141183460469231731687303715884105728n);
+    assert.strictEqual(wasm.optional_i128_identity(wasm.optional_i128_max()), 170141183460469231731687303715884105727n);
+    assert.strictEqual(wasm.optional_i128_identity(0x3_1415_9265_3598_7932_3846n), 0x3_1415_9265_3598_7932_3846n);
```
Comment (daxpedda) [inline tests/wasm/optional_primitives.js:113]: Whats that number? Just random?

**Review by daxpedda — CHANGES_REQUESTED** [review]
Great, thank you!

Just some nits and a changelog entry missing.

File: tests/wasm/optional_primitives.js (hunk 3 above)
Comment (RunDevelopment) [reply tests/wasm/optional_primitives.js:113]: It's pi. I just wanted to test a large random number.

File: tests/wasm/optional_primitives.js (hunk 2 above)
Comment (RunDevelopment) [reply tests/wasm/optional_primitives.js:120]: Good point.Hex would be a bit easier to understand here.

File: tests/wasm/optional_primitives.js (hunk 3 above)
Comment (daxpedda) [reply tests/wasm/optional_primitives.js:113]: Oh, right!

**Review by daxpedda — APPROVED** [review]
Thanks!

**Pull request merged**